sql-plugin/src/main/scala/com/nvidia/spark/rapids/BatchWithPartitionData.scala

```diff
@@ -0,0 +1,356 @@
+import com.nvidia.spark.rapids.Arm.{closeOnExcept, withResource}
+import com.nvidia.spark.rapids.RapidsPluginImplicits._
+
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.types.{StringType, StructType}
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+/**
+ * Wrapper class that specifies how many rows to replicate
+ * the partition value.
+ */
+case class PartitionRowData(rowValue: InternalRow, rowNum: Long)
+
+/**
+ * Class to wrap columnar batch and partition rows data and utility functions to merge them.
+ *
+ * @param inputBatch           Input ColumnarBatch.
+ * @param partitionedRowsData  Array of (row counts, InternalRow) pairs. Each entry specifies how
+ *                             many rows to replicate the partition value.
+ * @param partitionSchema      Schema of the partitioned data.
+ */
+class BatchWithPartitionData(
+    inputBatch: SpillableColumnarBatch,
+    partitionedRowsData: Array[PartitionRowData],
+    partitionSchema: StructType) extends AutoCloseable {
+
+  /**
+   * Merges the partitioned data with the input ColumnarBatch.
+   *
+   * @return Merged ColumnarBatch.
+   */
+  def mergeWithPartitionData(): ColumnarBatch = {
+    withResource(inputBatch.getColumnarBatch()) { columnarBatch =>
+      if (partitionSchema.isEmpty) {
+        columnarBatch // No partition columns, return the original batch.
+      } else {
+        val partitionColumns = buildPartitionColumns()
+        val numRows = partitionColumns.head.getRowCount.toInt
+        withResource(new ColumnarBatch(partitionColumns.toArray, numRows)) { partBatch =>
+          GpuColumnVector.combineColumns(columnarBatch, partBatch)
+        }
+      }
+    }
+  }
+
+  /**
+   * Builds GPU-based partition columns by mapping partitioned data to a sequence of
+   * GpuColumnVectors.
+   *
+   * @return A sequence of GpuColumnVectors, one for each partitioned column in the schema.
+   */
+  private def buildPartitionColumns(): Seq[GpuColumnVector] = {
+    closeOnExcept(new Array[GpuColumnVector](partitionSchema.length)) { partitionColumns =>
+      for ((field, colIndex) <- partitionSchema.zipWithIndex) {
+        val dataType = field.dataType
+        // Create an array to hold the individual columns for each partition.
+        val singlePartCols = partitionedRowsData.safeMap {
+          case PartitionRowData(valueRow, rowNum) =>
+            val singleValue = valueRow.get(colIndex, dataType)
+            withResource(GpuScalar.from(singleValue, dataType)) { singleScalar =>
+              // Create a column vector from the GPU scalar, associated with the row number.
+              ColumnVector.fromScalar(singleScalar, rowNum.toInt)
+            }
+        }
+
+        // Concatenate the individual columns into a single column vector.
+        val concatenatedColumn = if (singlePartCols.length > 1) {
+          ColumnVector.concatenate(singlePartCols: _*)
+        } else {
+          singlePartCols.head
+        }
```

Comment:
singlePartCols will leak, need something like:
```suggestion
        val concatenatedColumn = withResource(singlePartCols) { _ =>
          if (singlePartCols.length > 1) {
            ColumnVector.concatenate(singlePartCols: _*)
          } else {
            singlePartCols.head.incRefCount()
          }
        }
```